### ISSUE TITLE
[distributed] add a "small" transformer model to the test suite for testing usage

### ISSUE THREAD

**Issue by wanchaol** [pytorch/pytorch#115309]
We can add a "small" Transformer model to our test suite to test out our existing FSDP/TP/SP APIs to ensure no future breakge and give easy to read example about how to apply our APIs directly in tests.

**Comment by wconstab** [comment]
is this different from the tests here?
https://github.com/pytorch/pytorch/blob/main/test/distributed/test_dynamo_distributed.py#L178

**Comment by awgu** [comment]
Let me know if this lands. I have a fork of nanoGPT in the per-parameter-sharding FSDP PR, including inlined `torch.utils.checkpoint` as an option (to test that).

I have found the nanoGPT model to be really good for testing FSDP.

**Comment by wanchaol** [comment]
> is this different from the tests here? https://github.com/pytorch/pytorch/blob/main/test/distributed/test_dynamo_distributed.py#L178

@wconstab Yeah I was thinking about having a "small" (in term of model param size) model with model definition in our test suite, so instead of relying on the transformer package where it hides the model definition and hard to modify, a small toy transformer model could let us test things in a more detailed way, i.e. we can test meta init API or how the model get sharded/splitted with more details as we have the model definition

**Comment by tianyu-l** [comment]
> Let me know if this lands. I have a fork of nanoGPT in the per-parameter-sharding FSDP PR, including inlined `torch.utils.checkpoint` as an option (to test that).
> 
> I have found the nanoGPT model to be really good for testing FSDP.
> 
> Existing FSDP has `TransformerWithSharedParams`:
> 
> https://github.com/pytorch/pytorch/blob/bb7746275c1151a2d2113b2a3f323b3c599ff04b/torch/testing/_internal/common_fsdp.py#L212

@awgu Hey Andrew, just FYI the PR landed. Next I should learn how to incorporate FSDP into the test.
